nowcasting_dataset/manager.py

```diff
@@ -0,0 +1,384 @@
+                " Are you sure you mean to do this?!"
+            )
+
+        # Get the intersection of t0 time periods from all data sources.
+        t0_time_periods_for_all_data_sources = []
+        for data_source_name, data_source in self.data_sources.items():
+            logger.debug(f"Getting t0 time periods for {data_source_name}")
+            try:
+                t0_time_periods = data_source.get_contiguous_t0_time_periods()
+            except NotImplementedError:
+                # Skip data_sources with no concept of time.
+                logger.debug(f"Skipping {data_source_name} because it has not concept of datetime.")
+            else:
+                t0_time_periods_for_all_data_sources.append(t0_time_periods)
+
+        intersection_of_t0_time_periods = nd_time.intersection_of_multiple_dataframes_of_periods(
+            t0_time_periods_for_all_data_sources
+        )
+
+        t0_datetimes = nd_time.time_periods_to_datetime_index(
+            time_periods=intersection_of_t0_time_periods, freq=freq
+        )
+        logger.debug(
+            f"Found {len(t0_datetimes):,d} datetimes at freq=`{freq}` across"
+            f" DataSources={list(self.data_sources.keys())}."
+            f"  From {t0_datetimes[0]} to {t0_datetimes[-1]}."
+        )
+        return t0_datetimes
+
+    def sample_spatial_and_temporal_locations_for_examples(
+        self, t0_datetimes: pd.DatetimeIndex, n_examples: int
+    ) -> pd.DataFrame:
+        """
+        Computes the geospatial and temporal locations for each training example.
+
+        The first data_source in this DataSourceList defines the geospatial locations of
+        each example.
+
+        Args:
+            t0_datetimes: All available t0 datetimes.  Can be computed with
+                `DataSourceList.get_t0_datetimes_across_all_data_sources()`
+            n_examples: The number of examples requested.
+
+        Returns:
+            Each row of each the DataFrame specifies the position of each example, using
+            columns: 't0_datetime_UTC', 'x_center_OSGB', 'y_center_OSGB'.
+        """
+        shuffled_t0_datetimes = np.random.choice(t0_datetimes, size=n_examples)
+        # TODO: Speed this up by splitting the shuffled_t0_datetimes across multiple processors.
+        (
+            x_locations,
+            y_locations,
+        ) = self.data_source_which_defines_geospatial_locations.get_locations(shuffled_t0_datetimes)
+        return pd.DataFrame(
+            {
+                "t0_datetime_UTC": shuffled_t0_datetimes,
+                "x_center_OSGB": x_locations,
+                "y_center_OSGB": y_locations,
+            }
+        )
+
+    def _get_first_batches_to_create(
+        self, overwrite_batches: bool
+    ) -> dict[split.SplitName, dict[str, int]]:
+        """For each SplitName & for each DataSource name, return the first batch ID to create."""
```

Comment:
could you give an example of this return object? My head got a bit confused here

Reply:
Good point, it is a bit head-spin-making!  I've added an example.